src/AzModuleInstaller.ts

```diff
@@ -0,0 +1,159 @@
+            this.installResult = {
+                isInstalled: true,
+                moduleSource: "hostedAgentFolder"
+            };
+        }
+    }
+
+    private async tryInstallFromFolder() {
+        if (this.installResult.isInstalled) {
+            return;
+        }
+
+        if (FileUtils.pathExists(this.modulePath)) {
+            core.debug(`Az ${this.version} present at ${this.modulePath} as folder.`);
+            this.installResult = {
+                isInstalled: true,
+                moduleSource: "hostedAgentFolder"
+            };
+        }
+    }
+
+    private async tryInstallFromZip() {
+        if (this.installResult.isInstalled) {
+            return;
+        }
+
+        if (FileUtils.pathExists(this.moduleZipPath)) {
+            core.debug(`Az ${this.version} present at ${this.moduleZipPath} as zip, expanding it.`);
+            await new ArchiveTools(this.isWin).unzip(this.moduleZipPath, this.moduleContainerPath);
+            await FileUtils.deleteFile(this.moduleZipPath);
+            this.installResult = {
+                isInstalled: true,
+                moduleSource: "hostedAgentZip"
+            };
+        }
+    }
+
+    private async tryInstallFromGHRelease() {
+        if (this.installResult.isInstalled) {
+            return;
+        }
+
+        try {
+            const downloadUrl = await this.getDownloadUrlFromGHRelease();
+            core.debug(`Downloading Az ${this.version} from GHRelease using url ${downloadUrl}`);
+            await tc.downloadTool(downloadUrl, this.moduleZipPath, this.githubAuth);
+            core.debug(`Expanding Az ${this.version} downloaded at ${this.moduleZipPath} as zip.`);
+            await new ArchiveTools(this.isWin).unzip(this.moduleZipPath, this.moduleContainerPath);
+            await FileUtils.deleteFile(this.moduleZipPath);
+            this.installResult = {
+                isInstalled: true,
+                moduleSource: "hostedAgentGHRelease"
+            };
+        } catch (err) {
+            core.debug(err);
+            console.log("Download from GHRelease failed, will fallback to PSGallery");
+        }
+    }
+
+    private async tryInstallFromPSGallery() {
+        if (this.installResult.isInstalled) {
+            return;
+        }
+
+        await Utils.saveAzModule(this.version, this.modulePath);
+        this.installResult = {
+            isInstalled: true,
+            moduleSource: "hostedAgentPSGallery"
+        };
+    }
+
+    private async getDownloadUrlFromGHRelease() {
+        core.debug("Getting versions manifest from GHRelease.");
+        const releases = await tc.getManifestFromRepo(
+            "Azure",
+            "az-ps-module-versions",
+            this.githubAuth,
+            "main");
+        core.debug(JSON.stringify(releases));
+        const releaseInfo = releases.filter(release => release.version === this.version)?.[0];
```

Comment:
if someone gives a non existent version of azure powershell, then won't this logic send back null ? in that case, we should check for it and maybe throw appropriately

Reply:
added throw as appropriate